crates/server/src/server/server_future.rs

```diff
@@ -642,71 +673,146 @@ impl<T: RequestHandler> ServerFuture<T> {
+                    // TODO: need to consider timeout of total connect...
+                    let result =
+                        quic_handler(handler, streams, src_addr, dns_hostname, drain.clone()).await;
 
-                        if let Err(e) = result {
-                            warn!("quic stream processing failed from {src_addr}: {e}")
-                        }
-                    });
-
-                    reap_tasks(&mut inner_join_set);
-                }
+                    if let Err(e) = result {
+                        warn!("quic stream processing failed from {src_addr}: {e}")
+                    }
+                });
             }
+
+            reap_tasks(&mut inner_join_set);
+            Ok(())
         });
 
         Ok(())
     }
 
-    /// This will run until a background task of the trust_dns_server ends.
+    /// Returns the drain handle used for initiating the graceful
+    /// shutdown.
+    pub fn get_drain(&self) -> Drain {
+        self.drain.clone()
+    }
+
+    /// Triggers a graceful shutdown the server. All background tasks will stop accepting
+    /// new connections and the returned future will complete once all tasks have terminated.
+    ///
+    /// This is equivalent to calling [Drain::drain] followed by [Self::block_until_done].
+    pub fn shutdown_gracefully(self) -> impl Future<Output = Result<(), ProtoError>> {
+        self.drain.drain();
+        self.block_until_done()
+    }
+
+    /// This will run until all background tasks complete. If one or more tasks return an error,
+    /// one will be chosen as the returned error for this future.
     pub async fn block_until_done(mut self) -> Result<(), ProtoError> {
-        let result = self.join_set.join_next().await;
+        if self.join_set.is_empty() {
+            warn!("block_until_done called with no pending tasks");
+            return Ok(());
+        }
 
-        match result {
-            None => {
-                tracing::warn!("block_until_done called with no pending tasks");
-                Ok(())
+        // Now wait for all of the tasks to complete.
+        let mut out = Ok(());
+        while let Some(join_result) = self.join_set.join_next().await {
+            match join_result {
+                Ok(result) => {
+                    match result {
+                        Ok(_) => (),
+                        Err(e) => {
+                            // Save the last error.
+                            out = Err(e);
+                        }
+                    }
+                }
+                Err(e) => return Err(ProtoError::from(format!("Internal error in spawn: {e}"))),
             }
-            Some(Ok(x)) => x,
-            Some(Err(e)) => Err(ProtoError::from(format!("Internal error in spawn: {e}"))),
+        }
+        out
+    }
+}
+/// A Future that is used to signal server threads to begin a graceful shutdown by
```

Comment:
I think we should just skip this layer of abstraction, and just pass the `Arc<AtomicBool>` in directly. I also don't think the `Drain` naming is great in this context, I'd probably call it `shutting_down`.

Reply:
This abstraction being a future allows us to use it in situtations like `take_until`, which is used for UDP. I also think the abstraction is pretty light, and doesn't complicate the API.

Regarding the naming, the term `drain` is a pretty standard term referring to the trigger that begins a graceful shutdown. Both hyper and linkerd mix `drain` and `graceful` throughout their API and docs. Calling `drain` is more concise than calling `start_graceful_shutdown`. I added a method `shutdown_gracefully` as an all-in-one helper that both starts the graceful shutdown and returns the waiting future.